components/sidebar/sidebar_channel/sidebar_base_channel/sidebar_base_channel.tsx

```diff
@@ -2,25 +2,92 @@
 
+import {trackEvent} from 'actions/diagnostics_actions';
+import * as GlobalActions from 'actions/global_actions';
+
+import GlobeIcon from 'components/widgets/icons/globe_icon';
+import LockIcon from 'components/widgets/icons/lock_icon';
+
+import SidebarChannelLink from 'components/sidebar/sidebar_channel/sidebar_channel_link';
+import {localizeMessage} from 'utils/utils';
+import Constants from 'utils/constants';
+
 type Props = {
     channel: Channel;
     currentTeamName: string;
+    enableXToLeaveChannelsFromLHS?: string;
+    actions: {
+        leaveChannel: (channelId: any) => void;
+    };
 };
 
 type State = {
 
 };
 
 export default class SidebarBaseChannel extends React.PureComponent<Props, State> {
+    handleLeavePublicChannel = () => {
+        this.props.actions.leaveChannel(this.props.channel.id);
+        trackEvent('ui', 'ui_public_channel_x_button_clicked');
+    }
+
+    handleLeavePrivateChannel = () => {
+        GlobalActions.showLeavePrivateChannelModal({id: this.props.channel.id, display_name: this.props.channel.display_name});
+        trackEvent('ui', 'ui_private_channel_x_button_clicked');
+    }
+
+    getCloseHandler = () => {
+        const {channel, enableXToLeaveChannelsFromLHS} = this.props;
+
+        if (enableXToLeaveChannelsFromLHS === 'true') {
+            if (channel.type === Constants.OPEN_CHANNEL && channel.name !== Constants.DEFAULT_CHANNEL) {
+                return this.handleLeavePublicChannel;
+            } else if (channel.type === Constants.PRIVATE_CHANNEL) {
+                return this.handleLeavePrivateChannel;
+            }
+        }
+
+        return null;
+    }
+
+    getIcon = () => {
+        const {channel} = this.props;
+
+        if (channel.type === Constants.OPEN_CHANNEL) {
+            return (
+                <GlobeIcon className='icon icon__globe'/>
+            );
+        } else if (channel.type === Constants.PRIVATE_CHANNEL) {
+            return (
+                <LockIcon className='icon icon__lock'/>
+            );
+        }
+
+        return null;
+    }
+
     render() {
         const {channel, currentTeamName} = this.props;
 
+        let ariaLabelPrefix;
+        if (channel.type === Constants.OPEN_CHANNEL) {
+            ariaLabelPrefix = localizeMessage('accessibility.sidebar.types.public', 'public channel');
+        } else if (channel.type === Constants.PRIVATE_CHANNEL) {
+            ariaLabelPrefix = localizeMessage('accessibility.sidebar.types.private', 'private channel');
+        }
+
         return (
-            <Link to={`/${currentTeamName}/channels/${channel.name}`}>{channel.display_name}</Link>
+            <SidebarChannelLink
+                channel={channel}
+                link={`/${currentTeamName}/channels/${channel.name}`}
+                label={channel.display_name}
+                ariaLabelPrefix={ariaLabelPrefix}
+                closeHandler={this.getCloseHandler()!}
+                icon={this.getIcon()!}
```

Comment:
They are, but when TypeScript defines a type as optional it's really returning the type as `T | undefined`.

Since we have an eslint rule that explicitly banned the use of `undefined`, we have `strictNullChecks` on, and the method to determine the value of `getCloseHandler()` has to return a null, we need the `!` there for the TypeScript compiler not to complain.

This is something we need to address as part of our TypeScript conversion, which is outside the scope of this PR.

Reply:
I filed https://mattermost.atlassian.net/browse/MM-22672 to chat more about this later on.